src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+#ifndef CEREAL_ARCHIVES_BSON_HPP_
+#define CEREAL_ARCHIVES_BSON_HPP_
+
+#include <stack>
+#include <string>
+#include <vector>
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/types/value.hpp>
+
+namespace cereal {
+/**
+ * A templated stuct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type.
+ */
+template <class BsonT>
+struct is_bson {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_double>::value ||
```

Comment:
Added the UnderlyingBSONDataBase class. When a serializable class inherits from this base, deserialization will automatically keep track of the memory associated with the class's bsoncxx view types in a shared_ptr.